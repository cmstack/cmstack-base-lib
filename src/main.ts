import {
  aws_ec2 as ec2,
  aws_ecs as ecs,
  aws_ecs_patterns as patterns,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';


export interface BaseProps {
  readonly vpc?: ec2.IVpc;
}

// create a Base construct
export class Base extends Construct {
  readonly vpc: ec2.IVpc;
  constructor(scope: Construct, id: string, props?: BaseProps) {
    super(scope, id);

    this.vpc = props?.vpc ?? new ec2.Vpc(this, 'Vpc', { natGateways: 1 });

    // create alb farget service
    new patterns.ApplicationLoadBalancedFargateService(this, 'Fargate', {
      vpc: this.vpc,
      cpu: 512,
      memoryLimitMiB: 1024,
      taskImageOptions: {
        image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
        containerPort: 80,
        enableLogging: true,
      },
    });
  }
}