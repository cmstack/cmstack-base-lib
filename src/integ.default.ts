import {
  App,
  Stack,
  aws_ec2 as ec2,
} from 'aws-cdk-lib';

import { Base } from './main';

const env = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const stack = new Stack(app, 'BaseStack', { env });


const vpc = ec2.Vpc.fromLookup(stack, 'Vpc', { isDefault: true });

new Base(stack, 'Base', {
  vpc,
});