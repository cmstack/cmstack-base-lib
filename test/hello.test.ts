import {
  assertions,
  App,
  Stack,
} from 'aws-cdk-lib';
import { Base } from '../src';


test('match snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'BaseStack');
  new Base(stack, 'Base');
  const templace = assertions.Template.fromStack(stack);
  expect(templace).toMatchSnapshot();
});


// make sure the stack has 1 alb
test('alb', () => {
  const app = new App();
  const stack = new Stack(app, 'BaseStack');
  new Base(stack, 'Base');
  const templace = assertions.Template.fromStack(stack);
  templace.resourceCountIs('AWS::ElasticLoadBalancingV2::LoadBalancer', 1);
});