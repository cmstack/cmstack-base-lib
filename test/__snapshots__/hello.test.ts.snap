// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
{
  "Outputs": {
    "BaseFargateLoadBalancerDNSF37EC0D5": {
      "Value": {
        "Fn::GetAtt": [
          "BaseFargateLB5746C7C5",
          "DNSName",
        ],
      },
    },
    "BaseFargateServiceURLA5158BD8": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "http://",
            {
              "Fn::GetAtt": [
                "BaseFargateLB5746C7C5",
                "DNSName",
              ],
            },
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BaseFargateLB5746C7C5": {
      "DependsOn": [
        "BaseVpcPublicSubnet1DefaultRoute69EEA2AB",
        "BaseVpcPublicSubnet2DefaultRoute8645B21A",
      ],
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "BaseFargateLBSecurityGroupE0AED962",
              "GroupId",
            ],
          },
        ],
        "Subnets": [
          {
            "Ref": "BaseVpcPublicSubnet1Subnet04A08E07",
          },
          {
            "Ref": "BaseVpcPublicSubnet2Subnet92BCD4F8",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "BaseFargateLBPublicListenerECSGroup772F8809": {
      "Properties": {
        "Port": 80,
        "Protocol": "HTTP",
        "TargetGroupAttributes": [
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "ip",
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "BaseFargateLBPublicListenerF29E6C3D": {
      "Properties": {
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "BaseFargateLBPublicListenerECSGroup772F8809",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "BaseFargateLB5746C7C5",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "BaseFargateLBSecurityGroupE0AED962": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB BaseStackBaseFargateLBEEA7629F",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "BaseFargateLBSecurityGrouptoBaseStackBaseFargateServiceSecurityGroup95BD261E80C6F05D71": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "BaseFargateServiceSecurityGroup5E5A4F6C",
            "GroupId",
          ],
        },
        "FromPort": 80,
        "GroupId": {
          "Fn::GetAtt": [
            "BaseFargateLBSecurityGroupE0AED962",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "BaseFargateService76552E1E": {
      "DependsOn": [
        "BaseFargateLBPublicListenerECSGroup772F8809",
        "BaseFargateLBPublicListenerF29E6C3D",
      ],
      "Properties": {
        "Cluster": {
          "Ref": "EcsDefaultClusterMnL3mNNYNVpc18E0451A",
        },
        "DeploymentConfiguration": {
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 50,
        },
        "EnableECSManagedTags": false,
        "HealthCheckGracePeriodSeconds": 60,
        "LaunchType": "FARGATE",
        "LoadBalancers": [
          {
            "ContainerName": "web",
            "ContainerPort": 80,
            "TargetGroupArn": {
              "Ref": "BaseFargateLBPublicListenerECSGroup772F8809",
            },
          },
        ],
        "NetworkConfiguration": {
          "AwsvpcConfiguration": {
            "AssignPublicIp": "DISABLED",
            "SecurityGroups": [
              {
                "Fn::GetAtt": [
                  "BaseFargateServiceSecurityGroup5E5A4F6C",
                  "GroupId",
                ],
              },
            ],
            "Subnets": [
              {
                "Ref": "BaseVpcPrivateSubnet1Subnet8DFC4571",
              },
              {
                "Ref": "BaseVpcPrivateSubnet2Subnet3B24C618",
              },
            ],
          },
        },
        "TaskDefinition": {
          "Ref": "BaseFargateTaskDef2988244D",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "BaseFargateServiceSecurityGroup5E5A4F6C": {
      "Properties": {
        "GroupDescription": "BaseStack/Base/Fargate/Service/SecurityGroup",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "BaseFargateServiceSecurityGroupfromBaseStackBaseFargateLBSecurityGroup9BD54C9B80FCF870B0": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 80,
        "GroupId": {
          "Fn::GetAtt": [
            "BaseFargateServiceSecurityGroup5E5A4F6C",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "BaseFargateLBSecurityGroupE0AED962",
            "GroupId",
          ],
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "BaseFargateTaskDef2988244D": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Essential": true,
            "Image": "amazon/amazon-ecs-sample",
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "BaseFargateTaskDefwebLogGroup83EECD0B",
                },
                "awslogs-region": {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "Fargate",
              },
            },
            "Name": "web",
            "PortMappings": [
              {
                "ContainerPort": 80,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "Cpu": "512",
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "BaseFargateTaskDefExecutionRoleB55766C9",
            "Arn",
          ],
        },
        "Family": "BaseStackBaseFargateTaskDef429F498A",
        "Memory": "1024",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "BaseFargateTaskDefTaskRole0E044446",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "BaseFargateTaskDefExecutionRoleB55766C9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BaseFargateTaskDefExecutionRoleDefaultPolicy68D4A883": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BaseFargateTaskDefwebLogGroup83EECD0B",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BaseFargateTaskDefExecutionRoleDefaultPolicy68D4A883",
        "Roles": [
          {
            "Ref": "BaseFargateTaskDefExecutionRoleB55766C9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BaseFargateTaskDefTaskRole0E044446": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BaseFargateTaskDefwebLogGroup83EECD0B": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "BaseVpc4D0293C1": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "BaseVpcIGWD7707A1B": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "BaseVpcPrivateSubnet1DefaultRoute99EA3ACC": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "BaseVpcPublicSubnet1NATGateway152D82D5",
        },
        "RouteTableId": {
          "Ref": "BaseVpcPrivateSubnet1RouteTableFBF5C3D7",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "BaseVpcPrivateSubnet1RouteTableAssociation9AD9A06A": {
      "Properties": {
        "RouteTableId": {
          "Ref": "BaseVpcPrivateSubnet1RouteTableFBF5C3D7",
        },
        "SubnetId": {
          "Ref": "BaseVpcPrivateSubnet1Subnet8DFC4571",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "BaseVpcPrivateSubnet1RouteTableFBF5C3D7": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "BaseVpcPrivateSubnet1Subnet8DFC4571": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "BaseVpcPrivateSubnet2DefaultRouteABC4ECF4": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "BaseVpcPublicSubnet1NATGateway152D82D5",
        },
        "RouteTableId": {
          "Ref": "BaseVpcPrivateSubnet2RouteTable31D4362E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "BaseVpcPrivateSubnet2RouteTable31D4362E": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "BaseVpcPrivateSubnet2RouteTableAssociation2DCF8F61": {
      "Properties": {
        "RouteTableId": {
          "Ref": "BaseVpcPrivateSubnet2RouteTable31D4362E",
        },
        "SubnetId": {
          "Ref": "BaseVpcPrivateSubnet2Subnet3B24C618",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "BaseVpcPrivateSubnet2Subnet3B24C618": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "BaseVpcPublicSubnet1DefaultRoute69EEA2AB": {
      "DependsOn": [
        "BaseVpcVPCGW2058C1D6",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "BaseVpcIGWD7707A1B",
        },
        "RouteTableId": {
          "Ref": "BaseVpcPublicSubnet1RouteTable910E048A",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "BaseVpcPublicSubnet1EIP0FC17686": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "BaseVpcPublicSubnet1NATGateway152D82D5": {
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "BaseVpcPublicSubnet1EIP0FC17686",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "BaseVpcPublicSubnet1Subnet04A08E07",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "BaseVpcPublicSubnet1RouteTable910E048A": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "BaseVpcPublicSubnet1RouteTableAssociation9710265F": {
      "Properties": {
        "RouteTableId": {
          "Ref": "BaseVpcPublicSubnet1RouteTable910E048A",
        },
        "SubnetId": {
          "Ref": "BaseVpcPublicSubnet1Subnet04A08E07",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "BaseVpcPublicSubnet1Subnet04A08E07": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "BaseVpcPublicSubnet2DefaultRoute8645B21A": {
      "DependsOn": [
        "BaseVpcVPCGW2058C1D6",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "BaseVpcIGWD7707A1B",
        },
        "RouteTableId": {
          "Ref": "BaseVpcPublicSubnet2RouteTable7BFCE26C",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "BaseVpcPublicSubnet2RouteTable7BFCE26C": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "BaseVpcPublicSubnet2RouteTableAssociationADE2FB5B": {
      "Properties": {
        "RouteTableId": {
          "Ref": "BaseVpcPublicSubnet2RouteTable7BFCE26C",
        },
        "SubnetId": {
          "Ref": "BaseVpcPublicSubnet2Subnet92BCD4F8",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "BaseVpcPublicSubnet2Subnet92BCD4F8": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "BaseStack/Base/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "BaseVpcVPCGW2058C1D6": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "BaseVpcIGWD7707A1B",
        },
        "VpcId": {
          "Ref": "BaseVpc4D0293C1",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "EcsDefaultClusterMnL3mNNYNVpc18E0451A": {
      "Type": "AWS::ECS::Cluster",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
